ext/lcms2/utils/linkicc/linkicc.1
.\"Shiju P. Nair September 30, 2004
.\"Thomas Weber <tweber@debian.org> April 23, 2014
.TH LINKICC 1 "September 30, 2004"
.SH NAME
linkicc - little cms device link generator.
.SH SYNOPSIS
.B linkicc
.RI [ options ] " profiles"
.SH DESCRIPTION
lcms is a standalone CMM engine, which deals with the color management.
It implements a fast transformation between ICC profiles.
.B linkicc
is a little cms device link generator.
.P
Links two or more profiles into a single devicelink profile.
Colorspaces must be paired except Lab/XYZ, that can be interchanged.
.SH OPTIONS
.TP
.BR \-a\  NUM
Observer adaptation state (abs.col. only), (0..1.0, float value) [defaults to 1.0].
.TP
.B \-b
Black point compensation.
.TP
.BI \-c\  precision
Precision (0=LowRes, 1=Normal, 2=Hi-res) [defaults to 1].
.TP
.BI \-d\  description
Description text (quotes can be used).
.TP
.BI \-h\  NUM
Show summary of options and examples (0=help, 1=Built-in profiles, 2=Examples, 3=Contact information)
.TP
.BI \-k\  inklimit
Ink-limiting in % (CMYK only), (0..400.0, float value) [default 400.0].
.TP
.B \-l
Use linearization curves (may affect accuracy).
.TP
.BI \-n\  gridpoints
Alternate way to set precision, number of CLUT points.
.TP
.BI \-o\  profile
Output devicelink profile [defaults to 'devicelink.icm'].
.TP
.BI \-r\  profileversion
Profile version. (CAUTION: may change the profile implementation), (2.0..4.3, float value) [defaults to 4.3].
.TP
.BI \-t\  NUM
Rendering intent
.nf
.RS
0=Perceptual [default]
1=Relative colorimetric
2=Saturation
3=Absolute colorimetric
10=Perceptual preserving black ink
11=Relative colorimetric preserving black ink
12=Saturation preserving black ink
13=Perceptual preserving black plane
14=Relative colorimetric preserving black plane
15=Saturation preserving black plane
.RE
.fi
.TP
.BI \-v\  verbosity
Verbosity level, (0=None, 1=Normal, 2=High, 3=Very High) [defaults to 0].
.TP
.B \-x
Creatively, guess deviceclass of resulting profile.
.TP
.BI \-y\  copyright
Copyright notice (quotes can be used) ["No copyright, use freely"].
.TP
.B \-8
Creates 8-bit devicelink.
.SH BUILT-IN PROFILES
.nf
	*Lab2  -- D50-based v2 CIEL*a*b
	*Lab4  -- D50-based v4 CIEL*a*b
	*Lab   -- D50-based v4 CIEL*a*b
	*XYZ   -- CIE XYZ (PCS)
	*sRGB  -- sRGB color space
	*Gray22 - Monochrome of Gamma 2.2
	*Gray30 - Monochrome of Gamma 3.0
	*null   - Monochrome black for all input
	*Lin2222- CMYK linearization of gamma 2.2 on each channel
.fi
.SH EXAMPLES
.nf
To create 'devicelink.icm' from a.icc to b.icc:
	linkicc a.icc b.icc

To create 'out.icc' from sRGB to cmyk.icc:
	linkicc -o out.icc *sRGB cmyk.icc

To create a sRGB input profile working in Lab:
	linkicc -x -o sRGBLab.icc *sRGB *Lab

To create a XYZ -> sRGB output profile:
	linkicc -x -o sRGBLab.icc *XYZ *sRGB

To create a abstract profile doing softproof for cmyk.icc:
	linkicc -t1 -x -o softproof.icc *Lab cmyk.icc cmyk.icc *Lab

To create a 'grayer' sRGB input profile:
	linkicc -x -o grayer.icc *sRGB gray.icc gray.icc *Lab

To embed ink limiting into a cmyk output profile:
	linkicc -x -o cmyklimited.icc -k 250 cmyk.icc *Lab

.fi
.SH NOTES
For suggestions, comments, bug reports etc. send mail to
info@littlecms.com.
.SH SEE ALSO
.BR jpgicc (1),
.BR psicc (1),
.BR tificc (1),
.BR transicc (1)
.SH AUTHOR
This manual page was written by Shiju p. Nair <shiju.p@gmail.com>,
for the Debian project.


mupdf/CHANGES
List of changes since the last release

	api: Improved accessors for markup/ink/polygon annotation data.

	mutool run: Edit Markup, Ink, and Polygon annotation data.
	mutool run: Fill out form fields.

List of changes in MuPDF 1.16.1

	mupdf-x11: Changed key bindings to match mupdf-gl.
	mupdf-x11: Invert by luminance for dark mode.
	android: Fix compilation error.

List of changes in MuPDF 1.16.0

	api: Major overhaul of color management architecture.
	api: Improved functions to verify/sign PDF documents.
	api: Number tree accessor function pdf_lookup_number.
	api: Parse and handle more options for PWG output.
	api: Removed obsolete gproof document type.
	api: User callbacks for warning and error messages.

	epub: Changed default page size to A5.
	epub: Draw embedded SVG documents in EPUB/XHTML.
	epub: New Noto fonts covering more unicode scripts.
	epub: Support small-caps font-variant.
	pdf: Add Redact annotation type and function to apply redactions.
	pdf: Add/remove/change encryption and password when saving PDF files.
	pdf: Improvements to text handling in PDF filter processor.
	pdf: MP and DP operators now call begin/end_layer device methods.
	pdf: New and improved progressive loading.
	svg: Draw external images in SVG documents.

	mutool show: Add 'form' selector to list PDF form fields.
	mutool sign: Sign PDF documents from the command line.
	mutool sign: Verify signatures in PDF documents from the command line.
	viewer: Option to save a script of user actions for replay with mutool run.
	viewer: Runtime toggle for ICC and spot rendering.
	viewer: Tooltip when hovering over a link in mupdf-gl.

List of changes in MuPDF 1.15.0

General improvements:
	* WebAssembly build target and examples.
	* Improved forms API in both C and Java bindings.
	* Improved forms JavaScript support.
	* Create appearance streams for more form field types.
	* Fixed many bugs in ICC color management.
	* Fixed many memory leaks in error cleanup.
	* Fixed bugs in pdfwrite output.
	* Improved text extraction from LaTeX documents with math symbols.
	* Improved trace device formatting.
	* Support CBZ and CBT files larger than 2Gb.
	* Show table of contents for FB2 and XHTML documents.
	* Show embedded raster images in SVG and XHTML documents.
	* Show FB2 cover page.
	* Add option to save PDF files without encryption.
	* Add inhibit-spaces option to stext device to turn off missing space detection.
	* Simplified fz_try/fz_always/fz_catch macros.

mupdf-gl improvements:
	* Automatically open annotation editor when selecting an annotation.
	* Full page color tinting option in mupdf-gl.
	* Show/hide table of contents sections.
	* Trigger a reload with sighup.
	* Toggle spot color mode with 'e'.

mutool improvements:
	mutool show $PDF outline -- show outline with open/closed state indicator.
	mutool show $PDF js -- show document level javascript.
	mutool clean -A -- create appearance streams when missing.
	mutool clean -AA -- recreate all appearance streams.
	mutool run docs/examples/portfolio.js -- extract embedded files from PDF document.
	mutool run docs/examples/pdf-dejpx.js -- decompress JPEG2000 images in PDF document.
	mutool run docs/examples/fix-s22pdf.js -- fix fonts in documents generated by S22PDF.

Significant API changes:
	* New and improved forms API.
	* Renamed -DNO_ICC to -DFZ_ENABLE_ICC=0
	* Removed fz_annot superclass. Use pdf_annot and pdf_widget directly instead.
	* Annotations and Widgets now have separate enumerators.
	* Added optional separations argument to pixmap rendering utility functions.

Various cleanups:
	Per-function documentation comments moved to the source file.
	Hopefully they will stay up to date more often than when they
	were hidden away in the header file.

	Removed unused internal testing tools: mjsgen and jstest.

	Removed TGA output support. Use one of the more common and useful PAM,
	PNM, PNG, or TIFF formats instead.

	Removed support for PDF portfolios. This work-in-progress feature was
	never completed, and only worked for a small subset of files.

	Removed support for progressive loading. This rarely used feature added
	a lot of complexity and was an unending source of bugs and strange
	behaviour. Removing this feature has allowed us to clean up and
	simplify large amounts of code, and fix dozens of bugs in one fell
	swoop.

List of changes in MuPDF 1.14.0

* New features:
	* Added "Source Han Serif" CJK fallback font.
	* Added more scripts to the Noto fallback fonts.
	* Multi-page PNM support.
	* "mutool show" now supports a path syntax for selecting objects to show.

* Build system simplifications:
	* Auto-generated CMap, ICC, and JS source files are checked in to git.
	* Embedded CMap resources are now generated by a python script.
	* Embedded font resources are linked directly if using GNU ld or windows.
	* Namegen tool replaced by use of C macros.
	* Simplified Makefile.

* Annotation editing:
	* New annotation editing mode in mupdf-gl.
	* Can create, edit, and delete most annotation types.
	* Can create appearance streams for most annotation types.
	* Can create appearance streams for Tx form fields.
	* Can create appearance streams for Ch form fields.

* Form filling in mupdf-gl:
	* Can click buttons, checkboxes, and radioboxes.
	* Can fill out text fields using dialog box.
	* Can select choice options using dialog box.
	* Can verify and sign digital signatures.

* Improved UI for mupdf-gl:
	* Password dialog.
	* Error dialog.
	* Open/save file dialog.
	* Snap selection to words or lines by holding control or control+shift.
	* Save and restore current location, bookmarks, and navigation history.

* Bug fixes:
	* Improved CJK character fallback handling in EPUB.

* API changes:
	* Pass rectangle and matrix structs by value.
	* Replaced PDF_NAME_Xxx macros with PDF_NAME(Xxx).
	* Added PDF_TRUE, PDF_FALSE, and PDF_NULL constant pdf_obj* macros.
	* Added helper functions: pdf_dict_get_int, etc.
	* Removed 'doc' argument in pdf_new_int, etc.

ext/lcms2/utils/linkicc/linkicc.c
//---------------------------------------------------------------------------------
//
//  Little Color Management System
//  Copyright (c) 1998-2017 Marti Maria Saguer
//
// Permission is hereby granted, free of charge, to any person obtaining
// a copy of this software and associated documentation files (the "Software"),
// to deal in the Software without restriction, including without limitation
// the rights to use, copy, modify, merge, publish, distribute, sublicense,
// and/or sell copies of the Software, and to permit persons to whom the Software
// is furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
// THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//
//---------------------------------------------------------------------------------

#include "utils.h"

// ---------------------------------------------------------------------------------

static char* Description = "Devicelink profile";
static char* Copyright   = "No copyright, use freely";
static int   Intent = INTENT_PERCEPTUAL;
static char* cOutProf    = "devicelink.icc";
static int   PrecalcMode  = 1;
static int   NumOfGridPoints = 0;

static cmsFloat64Number ObserverAdaptationState = 1.0;  // According ICC 4.2 this is the default

static cmsBool BlackPointCompensation = FALSE;

static cmsFloat64Number InkLimit   = 400;
static cmsBool lUse8bits           = FALSE;
static cmsBool TagResult           = FALSE;
static cmsBool KeepLinearization   = FALSE;
static cmsFloat64Number Version    = 4.3;


// The manual
static
int Help(int level)
{
     switch (level) {

     default:
     case 0:

         fprintf(stderr, "\nlinkicc: Links profiles into a single devicelink.\n");

         fprintf(stderr, "\n");
         fprintf(stderr, "usage: linkicc [flags] <profiles>\n\n");
         fprintf(stderr, "flags:\n\n");
         fprintf(stderr, "%co<profile> - Output devicelink profile. [defaults to 'devicelink.icc']\n", SW);

         PrintRenderingIntents(NULL);

         fprintf(stderr, "%cc<0,1,2> - Precision (0=LowRes, 1=Normal, 2=Hi-res) [defaults to 1]\n", SW);
         fprintf(stderr, "%cn<gridpoints> - Alternate way to set precision, number of CLUT points\n", SW);
         fprintf(stderr, "%cd<description> - description text (quotes can be used)\n", SW);
         fprintf(stderr, "%cy<copyright> - copyright notice (quotes can be used)\n", SW);

         fprintf(stderr, "\n%ck<0..400> - Ink-limiting in %% (CMYK only)\n", SW);
         fprintf(stderr, "%c8 - Creates 8-bit devicelink\n", SW);
         fprintf(stderr, "%cx - Creatively, guess deviceclass of resulting profile.\n", SW);
         fprintf(stderr, "%cb - Black point compensation\n", SW);
         fprintf(stderr, "%ca<0..1> - Observer adaptation state (abs.col. only)\n\n", SW);
         fprintf(stderr, "%cl - Use linearization curves (may affect accuracy)\n", SW);
         fprintf(stderr, "%cr<v.r> - Profile version. (CAUTION: may change the profile implementation)\n", SW);
         fprintf(stderr, "\n");
         fprintf(stderr, "Colorspaces must be paired except Lab/XYZ, that can be interchanged.\n\n");

         fprintf(stderr, "%ch<0,1,2,3> - More help\n", SW);
         break;

     case 1:
         PrintBuiltins();
         break;

     case 2:

         fprintf(stderr, "\nExamples:\n\n"
             "To create 'devicelink.icm' from a.icc to b.icc:\n"
             "\tlinkicc a.icc b.icc\n\n"
             "To create 'out.icc' from sRGB to cmyk.icc:\n"
             "\tlinkicc -o out.icc *sRGB cmyk.icc\n\n"
             "To create a sRGB input profile working in Lab:\n"
             "\tlinkicc -x -o sRGBLab.icc *sRGB *Lab\n\n"
             "To create a XYZ -> sRGB output profile:\n"
             "\tlinkicc -x -o sRGBLab.icc *XYZ *sRGB\n\n"
             "To create a abstract profile doing softproof for cmyk.icc:\n"
             "\tlinkicc -t1 -x -o softproof.icc *Lab cmyk.icc cmyk.icc *Lab\n\n"
             "To create a 'grayer' sRGB input profile:\n"
             "\tlinkicc -x -o grayer.icc *sRGB gray.icc gray.icc *Lab\n\n"
             "To embed ink limiting into a cmyk output profile:\n"
             "\tlinkicc -x -o cmyklimited.icc -k 250 cmyk.icc *Lab\n\n");
         break;

     case 3:

         fprintf(stderr, "This program is intended to be a demo of the little cms\n"
             "engine. Both lcms and this program are freeware. You can\n"
             "obtain both in source code at http://www.littlecms.com\n"
             "For suggestions, comments, bug reports etc. send mail to\n"
             "info@littlecms.com\n\n");
    }

   exit(0);
}

// The toggles stuff
static
void HandleSwitches(int argc, char *argv[])
{
    int s;

    while ((s = xgetopt(argc,argv,"a:A:BbC:c:D:d:h:H:k:K:lLn:N:O:o:r:R:T:t:V:v:xX8y:Y:")) != EOF) {

    switch (s) {


        case 'a':
        case 'A':
            ObserverAdaptationState = atof(xoptarg);
            if (ObserverAdaptationState < 0 ||
                ObserverAdaptationState > 1.0)
                       FatalError("Adaptation state should be 0..1");
            break;

        case 'b':
        case 'B':
            BlackPointCompensation = TRUE;
           break;

        case 'c':
        case 'C':
            PrecalcMode = atoi(xoptarg);
            if (PrecalcMode < 0 || PrecalcMode > 2) {
                FatalError("Unknown precalc mode '%d'", PrecalcMode);
            }
           break;

       case 'd':
       case 'D':
           // Doing that is correct and safe: Description points to memory allocated in the command line.
           // same for Copyright and output devicelink.
           Description = xoptarg;
           break;

        case 'h':
        case 'H':
            Help(atoi(xoptarg));
            return;

        case 'k':
        case 'K':
            InkLimit = atof(xoptarg);
            if (InkLimit < 0.0 || InkLimit > 400.0) {
                FatalError("Ink limit must be 0%%..400%%");
            }
           break;


        case 'l':
        case 'L': KeepLinearization = TRUE;
           break;

       case 'n':
       case 'N':
           if (PrecalcMode != 1) {
               FatalError("Precalc mode already specified");
           }
           NumOfGridPoints = atoi(xoptarg);
           break;

        case 'o':
        case 'O':
            cOutProf = xoptarg;
           break;


       case 'r':
       case 'R':
          Version = atof(xoptarg);
          if (Version < 2.0 || Version > 4.3) {
              fprintf(stderr, "WARNING: lcms was not aware of this version, tag types may be wrong!\n");
          }
          break;

        case 't':
        case 'T':
            Intent = atoi(xoptarg);  // Will be validated latter on
            break;

        case 'V':
        case 'v':
            Verbose = atoi(xoptarg);
            if (Verbose < 0 || Verbose > 3) {
                FatalError("Unknown verbosity level '%d'", Verbose);
            }
            break;

        case '8':
            lUse8bits = TRUE;
            break;



        case 'y':
        case 'Y':
            Copyright = xoptarg;
            break;



       case 'x':
       case 'X': TagResult = TRUE;
           break;



       default:

           FatalError("Unknown option - run without args to see valid ones.\n");
        }
    }
}

// Set the copyright and description
static
cmsBool SetTextTags(cmsContext ContextID, cmsHPROFILE hProfile)
{
    cmsMLU *DescriptionMLU, *CopyrightMLU;
    cmsBool  rc = FALSE;

    DescriptionMLU  = cmsMLUalloc(ContextID, 1);
    CopyrightMLU    = cmsMLUalloc(ContextID, 1);

    if (DescriptionMLU == NULL || CopyrightMLU == NULL) goto Error;

    if (!cmsMLUsetASCII(ContextID, DescriptionMLU,  "en", "US", Description)) goto Error;
    if (!cmsMLUsetASCII(ContextID, CopyrightMLU,    "en", "US", Copyright)) goto Error;

    if (!cmsWriteTag(ContextID, hProfile, cmsSigProfileDescriptionTag,  DescriptionMLU)) goto Error;
    if (!cmsWriteTag(ContextID, hProfile, cmsSigCopyrightTag,           CopyrightMLU)) goto Error;

    rc = TRUE;

Error:

    if (DescriptionMLU)
        cmsMLUfree(ContextID, DescriptionMLU);
    if (CopyrightMLU)
        cmsMLUfree(ContextID, CopyrightMLU);
    return rc;
}



int main(int argc, char *argv[])
{
    int i, nargs, rc;
    cmsHPROFILE Profiles[257];
    cmsHPROFILE hProfile;
    cmsUInt32Number dwFlags;
    cmsHTRANSFORM hTransform = NULL;
    cmsContext ContextID = NULL;

    // Here we are
    fprintf(stderr, "little cms ICC device link generator - v2.2 [LittleCMS %2.2f]\n", LCMS_VERSION / 1000.0);
    fflush(stderr);

    // Initialize
    InitUtils(ContextID, "linkicc");
    rc = 0;

    // Get the options
    HandleSwitches(argc, argv);

    // How many profiles to link?
    nargs = (argc - xoptind);
    if (nargs < 1)
        return Help(0);

    if (nargs > 255) {
        FatalError("Holy profile! what are you trying to do with so many profiles!?");
        goto Cleanup;
    }

    // Open all profiles
    memset(Profiles, 0, sizeof(Profiles));
    for (i=0; i < nargs; i++) {

        Profiles[i] = OpenStockProfile(ContextID, argv[i + xoptind]);
        if (Profiles[i] == NULL) goto Cleanup;

        if (Verbose >= 1) {
            PrintProfileInformation(ContextID, Profiles[i]);
        }
    }

    // Ink limiting
    if (InkLimit != 400.0) {
        cmsColorSpaceSignature EndingColorSpace = cmsGetColorSpace(ContextID, Profiles[nargs-1]);
        Profiles[nargs++] = cmsCreateInkLimitingDeviceLink(ContextID, EndingColorSpace, InkLimit);
    }

    // Set the flags
    dwFlags = cmsFLAGS_KEEP_SEQUENCE;
    switch (PrecalcMode) {

        case 0: dwFlags |= cmsFLAGS_LOWRESPRECALC; break;
        case 2: dwFlags |= cmsFLAGS_HIGHRESPRECALC; break;
        case 1:
            if (NumOfGridPoints > 0)
                dwFlags |= cmsFLAGS_GRIDPOINTS(NumOfGridPoints);
            break;

        default:
            {
                FatalError("Unknown precalculation mode '%d'", PrecalcMode);
                goto Cleanup;
            }
    }

    if (BlackPointCompensation)
        dwFlags |= cmsFLAGS_BLACKPOINTCOMPENSATION;

    if (TagResult)
        dwFlags |= cmsFLAGS_GUESSDEVICECLASS;

    if (KeepLinearization)
        dwFlags |= cmsFLAGS_CLUT_PRE_LINEARIZATION|cmsFLAGS_CLUT_POST_LINEARIZATION;

    if (lUse8bits) dwFlags |= cmsFLAGS_8BITS_DEVICELINK;

     cmsSetAdaptationState(ContextID, ObserverAdaptationState);

    // Create the color transform. Specify 0 for the format is safe as the transform
    // is intended to be used only for the devicelink.
    hTransform = cmsCreateMultiprofileTransform(ContextID, Profiles, nargs, 0, 0, Intent, dwFlags|cmsFLAGS_NOOPTIMIZE);
    if (hTransform == NULL) {
        FatalError("Transform creation failed");
        goto Cleanup;
    }

    hProfile =  cmsTransform2DeviceLink(ContextID, hTransform, Version, dwFlags);
    if (hProfile == NULL) {
        FatalError("Devicelink creation failed");
        goto Cleanup;
    }

    SetTextTags(ContextID, hProfile);
    cmsSetHeaderRenderingIntent(ContextID, hProfile, Intent);

    if (cmsSaveProfileToFile(ContextID, hProfile, cOutProf)) {

        if (Verbose > 0)
            fprintf(stderr, "Ok");
    }
    else
        FatalError("Error saving file!");

    cmsCloseProfile(ContextID, hProfile);


Cleanup:

    if (hTransform != NULL) cmsDeleteTransform(ContextID, hTransform);
    for (i=0; i < nargs; i++) {

        if (Profiles[i] != NULL) cmsCloseProfile(ContextID, Profiles[i]);
    }

    return rc;
}
